### PR TITLE
Updates

### PR DESCRIPTION
**Pull request by abaire**
Lots of changes to bring master up to date with my latest. 

- Some major refactoring and automated style enforcement via `pre-commit`
- Lots of pylint fixes
- Added ability to skip cache and RDI dumps
- Support for producing a raw log of pgraph invocations
- Enables dumping of textures as well as surfaces
- Enables alpha preservation in dumps
- Reads textures through AGP to bypass tiling issues rather than attempting to disable tiling

Probably more things scattered in there as well but those are the highlights.

### CODE REVIEW

File: kick_fifo.asm [hunk 1]
```diff
@@ -2,49 +2,75 @@
```
Comment (abaire) [inline kick_fifo.asm:2]: It'd be good to take a detailed look at this change just to make sure I'm not doing something stupid

File: Texture.py [hunk 2]
```diff
@@ -1,163 +1,326 @@
...
+# FIXME: Move to xboxpy.NV2A.PGRAPH.Texture
 
 from collections import namedtuple
 from PIL import Image
 
+from Xbox import Xbox
+import XboxHelper
+from xboxpy import nv2a
+
+# Value that may be added to contiguous memory addresses to access as ADDR_AGPMEM, which
```
Comment (abaire) [inline Texture.py:18]: This is what I do in the pgraph test suite; relatively confident that it works correctly (since it does there, and so far has for all my test cases here), but if there are any known reasons not to take this approach please let me know.

File: kick_fifo.asm [hunk 3]
```diff
@@ -2,49 +2,75 @@
...
+%define DMA_PUSH_ADDR               0xFD003240
+%define DMA_PULL_ADDR               0xFD003244
+
 kick_fifo:
 
-; Load expected_put into reg
-mov edx, dword [esp+4]
+push ebp
+mov ebp, esp
+sub esp, 4
```
Comment (antangelo) [inline kick_fifo.asm:14]: What's the purpose of decrementing the stack pointer here? My x86 assembly is a bit rusty, so I'm probably just missing something

Comment (abaire) [reply kick_fifo.asm]: Cruft from my flailing around trying to get this to work on xemu as well as HW where I ran out of registers; removed

File: NV2ALog.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+
+class NV2ALog:
+    """Manages the nv2a log file."""
+
+    def __init__(self, path):
+        self.path = path
+
+        with open(self.path, "w", encoding="utf8") as logfile:
+            logfile.write("xemu style NV2A log from nv2a-trace.py\n\n")
```
Comment (JayFoxRox) [inline NV2ALog.py:16]: I still don't like xemu and don't agree that it should set standards for how tools should operate / format their output. At the very least, I'd rather not have it mentioned in this codebase - this tool is made for hardware, not for emulators like xemu.

In general, I'd suggest to split the display of the captured buffers into a separate step.

File: README.md [hunk 5]
```diff
@@ -51,6 +51,14 @@ It will automatically start tracing.
 **This tool may also (temporarily) corrupt the state of your Xbox.**
 If this tool does not work, please retry a couple of times.
 
+### Contributing
+
+This project uses [Black](https://pypi.org/project/black/) for automatic formatting.
+You can use the pre-commit in the `githooks` directory to reformat any changes on commit
```
Comment (JayFoxRox) [inline README.md:57]: I'd rather not have githooks as part of the repository (personal preference, and I found them annoying in projects that used them.

Comment (JayFoxRox) [inline README.md:57]: I'm also skeptical about auto-formatting, although I'm fine with code conventions in some standard format (I'm not sure what `black` is or wether it's a standard for Python, but I'd assume that some good standard exists for Python, given how over-specified the language is).

File: kick_fifo.asm [hunk 6]
```diff
@@ -2,49 +2,74 @@
...
+%define DMA_PUSH_ADDR               0xFD003240
+%define DMA_PULL_ADDR               0xFD003244
+
 kick_fifo:
 
-; Load expected_put into reg
-mov edx, dword [esp+4]
+push ebp
+mov ebp, esp
+push ebx
```
Comment (JayFoxRox) [inline kick_fifo.asm:14]: These prologues / epilogues should not be necessary because xboxpy `CALL` will store / restore all general-purpose registers.

File: kick_fifo.asm [hunk 7]
```diff
@@ -2,49 +2,74 @@
...
+mov eax, 0x32555359
 
-; Mark run as very bad if PUT changed
-cmp edx, dword [0xFD003240]
-je skip
+pause_pusher:
+; pause_fifo_pusher(xbox):
+;
+;state = xbox.read_u32(CACHE_PUSH_STATE)
+;xbox.write_u32(CACHE_PUSH_STATE, state & ~DMA_PUSH_ACCESS)
```
Comment (JayFoxRox) [inline kick_fifo.asm:59]: These old Python style comments should be removed in a cleanup, especially if you are adding named constants.

File: Texture.py (hunk 2 above)
Comment (JayFoxRox) [reply Texture.py:18]: Interesting hack. I'm surprised if this works, because to my knowledge, it isn't a feature of AGP, but of the GPU memory subsystem (as far as I remember, each tile can be configured with the surface format directly).
But if it works, then this is probably a good solution.

File: nv2a-trace.py [hunk 8]
```diff
@@ -19,7 +19,8 @@
 import Trace
 
 # pylint: disable=invalid-name
-_enable_experimental_disable_z_compression_and_tiling = True
+# TODO: Remove tiling suppression once AGP read in Texture.py is fully proven.
+_enable_experimental_disable_z_compression_and_tiling = False
```
Comment (JayFoxRox) [inline nv2a-trace.py:23]: I'd rather have this proven to work and then remove this code path entirely?
It makes the capturing much less invasive and some rather troublesome dead code could be emulated (seeing how capturing must be started twice, once to disable tiling, and then again to actually capture untiled memory once the GPU has actually disabled tiling on the next frames - some buffers might remain corrupted forever).

File: nv2a-trace.py [hunk 9]
```diff
@@ -257,6 +265,18 @@ def _parse_args():
...
 
+        parser.add_argument(
+            "--alpha-mode",
+            default="drop",
+            choices=["drop", "keep", "both"],
+            help=(
+                "Define how the alpha channel is handled in color graphical resources.\n"
+                "  drop: Discard the alpha channel\n"
+                "  keep: Save the alpha channel\n"
+                "  drop: Save two dumps, one with the alpha channel and one without\n"
```
Comment (JayFoxRox) [inline nv2a-trace.py:288]: Ideally, one would always dump all the channels and then fix the HTML with some simple JS (or SVG) to enable alpha changes in the HTML page (so it can show RGB-only / Alpha-only / Both active / Display individual channels / ...).
In general, I'd prefer if more of the texture decoding moved into the log viewer, so the original data can be preserved (which is interesting when looking at deferred rendering in games like Shrek, or stuff like HILO formats or RGBm for HDR in games like Wrecked).

File: XboxHelper.py [hunk 10]
```diff
@@ -289,21 +294,24 @@ def print_cache_state(self):
         print("Put / Pusher enabled: %s" % ("Yes" if (push_state & 1) else "No"))
         print("Get / Puller enabled: %s" % ("Yes" if (pull_state & 1) else "No"))
 
-        print("Cache:")
-        for i in range(128):
+        if print_contents:
+            print("Cache:")
+            # Each CACHE entry is a pair of 32-bit integers and there are 128 entries.
+            methods_and_data = self.xbox.read(CACHE1_METHOD, 8 * 128)
```
Comment (JayFoxRox) [inline XboxHelper.py:322]: This is risky, because it depends on the implementation of the `read`. You are accessing MMIO here.
I believe I specified xboxpy to load 32 bit chunks first, then 16 bit and finally 8 bit towards the end of the buffer.
So this *should* work, but another backend in xboxpy might still get it wrong, and a naive port to another backend would also fail.

I think it's better to keep this somewhat specific at least (so consider making a self-documenting high-level wrapper like `read_multiple_u32` / `read_u32(..., count=X)` or similar).

The main `read` operation in xboxpy might be optimized in the future. I had always had plans to enable some form of page-mirroring which actually mirrors the page-table and installs a watchdog to look for changes, so only touched changes would be transferred again, with the client keeping a mirror of the memory. Additionally, I had always planned to add compression for larger transfers, or DMA to blast the pages directly via ethernet (without lwip interaction).
xboxpy basically needs a way to differentiate MMIO and bulk-data-read/write.

The performance gain of this could be massive.

File: nv2a-trace.py [hunk 11]
```diff
@@ -66,7 +72,12 @@ def _wait_for_stable_push_buffer_state(
         # So we pause the pusher again to validate our state
         xbox_helper.pause_fifo_pusher()
 
-        time.sleep(1.0)
+        time.sleep(0.1)
```
Comment (JayFoxRox) [inline nv2a-trace.py:75]: Why is this possible?

I usually fine-tuned most of these waits after running into issues with various games.
I assume I had my reasons for choosing such high wait time.

Ideally none of these waits would exist, instead, there'd be code on the Xbox which actively monitors the GPU state (and also controls and guards it more reliably, against changes by the game).

File: Trace.py [hunk 12]
```diff
@@ -215,7 +215,7 @@ def run_fifo(self, put_target):
         self._write_put(put_target)
 
         # FIXME: we can avoid this read in some cases, as we should know where we are
-        self.real_dma_get_addr = self.xbox.read_u32(XboxHelper.DMA_GET_ADDR)
+        self.real_dma_get_addr = self.xbox.read_u32(XboxHelper.DMA_PULL_ADDR)
```
Comment (JayFoxRox) [inline Trace.py]: If this is wrong, then the variable name on the LHS is also broken

File: XboxHelper.py [hunk 13]
```diff
@@ -7,6 +7,9 @@
 import atexit
 import time
 
+# For general information on PFIFO, see
+# https://envytools.readthedocs.io/en/latest/hw/fifo/intro.html
```
Comment (JayFoxRox) [inline XboxHelper.py:23]: The naming by envytools / nouveau diverges from the nvidia names which were part of the Open-Source linux drivers and some reverse-engineered code.

Usually, the Xbox community sticks to the proprietary names due to the incompleteness of the nouveau names (although that sucks, because the proprietary names are hard to come by for most people. Those who have them usually whitelist those they are using on a case-by-case basis).

File: Trace.py [hunk 14]
```diff
@@ -64,20 +64,19 @@ def _dump_pfb(xbox):
...
-    xbox.write_u32(NV10_PGRAPH_RDI_INDEX, offset)
-    data = bytearray()
-    for _ in range(count):
-        word = xbox.read_u32(NV10_PGRAPH_RDI_DATA)
-        data += struct.pack("<L", word)
+    original_offset = ExchangeU32.exchange_u32(xbox, NV10_PGRAPH_RDI_INDEX, offset)
+
+    data = xbox.read(NV10_PGRAPH_RDI_DATA, count * 4)
+
+    xbox.write_u32(NV10_PGRAPH_RDI_INDEX, original_offset)
```
Comment (JayFoxRox) [inline Trace.py:69]: I'm not sure if it's even possible to restore the RDI index.
As far as I remember, the RDI has an internal state machine.

Reading the DATA register 4 times returns X,Y,Z,W (not in that order as far as I remember), but during that time the INDEX register will stay constant, only on the final read I think it auto-increments the INDEX.
If you touch the INDEX register, you may or may not be resetting the internal state machine.

File: Trace.py [hunk 15]
```diff
@@ -64,20 +64,19 @@ def _dump_pfb(xbox):
...
     NV10_PGRAPH_RDI_DATA = 0xFD400754
 
-    xbox.write_u32(NV10_PGRAPH_RDI_INDEX, offset)
-    data = bytearray()
-    for _ in range(count):
-        word = xbox.read_u32(NV10_PGRAPH_RDI_DATA)
-        data += struct.pack("<L", word)
+    original_offset = ExchangeU32.exchange_u32(xbox, NV10_PGRAPH_RDI_INDEX, offset)
+
+    data = xbox.read(NV10_PGRAPH_RDI_DATA, count * 4)
```
Comment (JayFoxRox) [inline Trace.py:67]: The `RDI_DATA` address is a single register, not a memory region.
It might occupy multiple addresses for this kind of bulk-reading, but then the `count` must be checked / verified or people might read incorrect data towards the end of the buffer (as they read beyond the `RDI_DATA` register).

File: Trace.py [hunk 16]
```diff
@@ -314,7 +314,16 @@ def _dump_texture(self, index):
...
         offset = self.xbox.read_u32(XboxHelper.PGRAPH_TEXOFFSET0 + reg_offset)
-        pitch = self.xbox.read_u32(XboxHelper.PGRAPH_TEXCTL1_0 + reg_offset) >> 16
+        # FIXME: read pitch from registers instead of guessing
+        # FIXME: clean up associated fallback code in Texture.py
+        # There is a case in Morrowind where pitch is set to 0x8 which is clearly wrong
+        #   Texture 0 [0x03CC6000, 128 x 128 x 1 (pitch: 0x8), format 0xC]
+        #   Texture 1 [0x03C98000, 64 x 64 x 1 (pitch: 0x8), format 0x6]
+        # Double check the register for correctness and see if swizzling has something
+        # to do with it (e.g., like how NV097_SET_TEXTURE_IMAGE_RECT is only needed by
+        # linear formats).
```
Comment (JayFoxRox) [inline Trace.py:324]: Not sure where the comment originated (is it mine?).
But pitch is only used for unswizzled formats - and it *is* used there.

File: Trace.py [hunk 17]
```diff
@@ -326,8 +335,8 @@ def _dump_texture(self, index):
         depth = 1 << depth_shift
 
         self._dbg_print(
-            "Texture %d [0x%08X, %d x %d x %d (pitch: 0x%X), format 0x%X]"
-            % (index, offset, width, height, depth, pitch, fmt_color)
+            "Texture %d [0x%08X, %d x %d x %d (pitch: 0x%X <ignored>), format 0x%X]"
```
Comment (JayFoxRox) [inline Trace.py:338]: This is incorrect

File: Trace.py [hunk 18]
```diff
@@ -326,8 +335,8 @@ def _dump_texture(self, index):
...
+            % (index, offset, width, height, depth, reg_pitch, fmt_color)
```
Comment (JayFoxRox) [inline Trace.py:339]: This must differentiate between swizzled and unswizzled; but usually I'd defer these sort of things to the decoding phase, after capturing.

I'd usually try to only decode what is necessary for capturing (say if you need to download a palette, or know the pixel-size from the format).
You should still log everything, because I found games which even controlled inactive registers as if they were being used - so having this additional data might be useful in some instances, even if it's not relevant for decoding.

**Review by abaire — COMMENTED** [review]
FYI, I suspect it's not worth retaining all of the commit history on this so I've just added a single commit responding to all outstanding comments. To make it easier to review just the delta, I'm refraining from squashing until things are approved.

File: Trace.py [hunk 19]
```diff
@@ -55,18 +55,28 @@ def _dump_pfb(xbox):
...
+    xbox.write_u32(NV10_PGRAPH_RDI_INDEX, offset)
+    data = bytearray()
+    for _ in range(count):
+        word = xbox.read_u32(NV10_PGRAPH_RDI_DATA)
+        data += struct.pack("<L", word)
+
+    # FIXME: Restore original RDI?
+    # Context from JFR: It may not be possible to restore the original index.
+    # If you touch the INDEX register, you may or may not be resetting the internal state
+    # machine.
```
Comment (JayFoxRox) [inline Trace.py:79]: Again: Please test instead of quoting.

File: NV2ALog.py (hunk 4 above)
Comment (JayFoxRox) [reply NV2ALog.py]: > I'm not sure I follow; do you mean store all the info currently written into the HTML and nv2a logs and then process into separate outputs?

One could just dump the raw pushbuffers in the capturing phase, doing minimal processing on them.
Then, in a second phase, decode these into HTML or other log formats.

I expect future improvements to the decoding phase, whereas the capturing phase likely won't change much.

During development of nv2a-trace, I had collected numerous traces of interesting scenes, but often they became outdated because of missing data in captures or the decoded portion.
By capturing more data and decoding later, I was able to improve my decoders without having to recapture a large amount of scenes.
So I think it's best to decouple log generation from the capturing / have a log format which stores as much information as possible (even if it might not be human readable).

File: Trace.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (JayFoxRox) [reply Trace.py]: > that is mostly the overhead of XBDM

Yes, there's a lot of networking overhead.
Both, for small transactions which are done individually, but also for large transfers which are very slow.

I had always planned to (eventually) run nv2a-trace through https://github.com/JayFoxRox/xbe-loader, nkpatcher or trainers (by integrating nxdk-rdt support, or some plugin system).
There has even been work to emulate the Xbox NIC so we have a real NIC (for the debugger) and a virtual NIC (for Xbox games to use).

The protocol of XBDM isn't ideal and the high level operations in xboxpy (which extend XBDM) are not ideal either.
For nv2a-trace, a hotfix has been to write some custom assembly for the most critical portions. Ideally, we'd be able to load C plugins into a running XBE, and we'd be able to do so without XBDM.

So ideally, nv2a-trace would be partially a C/ASM program running on the Xbox, partially a Python script for capturing on the PC side, and then Python (or Javascript / Typescript, due to HTML being so useful for portable UIs) for display.

Also note that I also worked on dxt support for nxdk for a while (starting with https://github.com/JayFoxRox/xbox-fps-overlay using https://github.com/JayFoxRox/nxdk/pulls?q=is%3Apr+is%3Aopen+dxt), so that was also an option I considered (although I didn't like how crashy the dxt-loader in the XDK was, and wanted to avoid proprietary stuff like the XDK).

Anyhow, I digress.

Long story short: Yes, XBDM is slow. And while we have options to make it fast, I'd ideally like to see a solution that is FOSS and runs on retail Xboxes and also works with multiplayer games (which often break with XBDM).

File: XboxHelper.py (hunk 10 above)
Comment (JayFoxRox) [reply XboxHelper.py]: **Reading multiple u32**

I think a smarter protocol which batches multiple operation would be useful.
Ideally there'd be something which disables interrupts, disables memory protection, then runs a number of queued commands, and only then enables protection and interrupts again.

Basically what we'd need is a transaction system in xboxpy, so you can batch multiple actions together.

**Memory mirroring to optimize bulk reads**

I'm not sure where to put some memory-mirroring system either. In theory, nothing prevents it from being independent of the xboxpy backend, so it could live in nv2a-trace.
The tricky bit is that it would probably have to install an interrupt handler.

This could be implemented in xboxpy, or in nv2a-trace, although I'd prefer it to live in nv2a-trace, because that's probably the only user.

---

The backends like XBDM and nxdk-rdt were meant for tiny atomic operations. The macro operations in nxdk-rdt or xbdm are useful for humans interacting with the Xbox, but they are pretty useless most of the time, while adding a lot of complexity.

Originally, xboxpy was meant to provide a high level view of the Xbox (on top of the atomic operations in the backends). Therefore xboxpy had an API for the APU and NV2A. However, it led to unnecessary complexity and most of the time there was a single user for those device specific APIs and then those APIs were often too complicated.
Instead, I feel like it was much more useful to be able to inject and call code manually from the actual applications / scripts.

So if I had to write xboxpy from scratch, I'd make it transaction based and focus on having a simple interface for MMIO, reads/writes and calls, potentially with the option to load C code (although I'd probably write the binary loader in Python on the PC side).

But I digress again.

File: XboxHelper.py [hunk 20]
```diff
@@ -246,14 +250,15 @@ def resume_fifo_pusher(self):
         if self.delay():
             pass
 
-    def allow_populate_fifo_cache(self):
+    def allow_populate_fifo_cache(self, wait_time=0.05):
```
[the remark this answers is not in the document]
Comment (JayFoxRox) [reply XboxHelper.py]: > I'm assuming you're talking about the `wait_time` specifically?

Yes.

> As far as I can tell the current xbdm transport layer is limited by having a timeout on hack invocations

I don't remember. But nv2a-trace clearly needs faster networking and more code running on the Xbox.
Usually I'd not care if capturing a scene took an hour or two as long as it captured all necessary detail.

I don't think I've ever had problems with the games time diverging during a capture, but I do remember crashes when returning from the capture, because the games internal timers diverged too much.

File: XboxHelper.py (hunk 13 above)
Comment (JayFoxRox) [reply XboxHelper.py:22]: I didn't look into specific registers (and don't have time for it), but I assume nv2a-trace already uses the XQEMU names?

I believe the XQEMU header should be using the nvidia names.

File: Texture.py (hunk 2 above)
Comment (JayFoxRox) [reply Texture.py:18]: Did you check the depth-buffers?
You probably also have to disable compression on those, but I don't remember where I do that.

Ideally we'd be able to decompress on the PC side, but I don't know how this works (but I know there's at least *some* information about it in tech articles; maybe even nouveau).

File: README.md (hunk 5 above)
Comment (JayFoxRox) [reply README.md:57]: Not a fan, but other maintainers should decide

Comment (JayFoxRox) [reply README.md:57]: Not a fan, but other maintainers should decide

File: Trace.py [hunk 21]
```diff
@@ -55,18 +55,28 @@ def _dump_pfb(xbox):
...
-    original_offset = ExchangeU32.exchange_u32(xbox, NV10_PGRAPH_RDI_INDEX, offset)
-
-    data = xbox.read(NV10_PGRAPH_RDI_DATA, count * 4)
-
-    xbox.write_u32(NV10_PGRAPH_RDI_INDEX, original_offset)
+    # JFR: Reading the DATA register 4 times returns X,Y,Z,W (not in that order as far
+    # as I remember), but during that time the INDEX register will stay constant, only
+    # on the final read I think it auto-increments the INDEX.
+    # It is not safe and likely incorrect to do a bulk read so this must be done
+    # individualy despite the interface communication overhead.
```
[the remark this answers is not in the document]
Comment (abaire) [reply Trace.py:69]: I can remove the attribution if you'd prefer (it does mean that there's a chance someone will come ask me via git blame and I'd send them to look at this PR if I didn't already know), but I am both reluctant to make this PR any more complicated than it already is (assuming after testing I do find some way to speed things up) and concerned about not leaving some sort of explanation behind. 

I think without a comment here there's a high chance that another contributor will follow in my footsteps; see that RDI dumping is so slow that unpredictable instability prevents capture of even a single frame, see that there's a tight loop instead of a bulk read, and erroneously attempt a bulk read to remove overhead.

As context, both myself and @antangelo have experienced transient failures that seem unpredictable and manifest as either socket timeouts or (more often) getmem failures in `xboxpy`. Running the same capture in as close as possible to the same state often leads to different results though, making this very hard to diagnose.
As further context, dumping a single frame of Morrowind (even with my broken RDI approach which sped up the read to the point of insignificance) took over three hours in the attempt that worked, with at least another hour on attempts that failed along the way.

My worry is that, without some warning, there will be a very strong drive to improve performance by folks who are way less knowledgeable about the internals than the maintainers and it'll be quite discouraging to contribution if they put in significant effort (e.g., if they don't already know Python or aren't programmers by trade) and learn after the fact that their change isn't actually viable.

File: Trace.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (abaire) [reply Trace.py]: This makes absolute sense to me. I actually tried following in your footsteps with dxt's when I was putting together https://github.com/abaire/xbdm_gdb_bridge but the nxdk had changed to the point that I wasn't able to get it working well. The fact that a bad dxt can prevent the box from booting far enough to allow deletion of the dxt is also a huge concern for me.

Being able to patch in a C program would be huge and is something I'd already put on my wishlist to investigate (also the ability to ensure they're loaded past 64 megs for modded boxes to deal with scenarios where most of the memory is spoken for, but perhaps that's already handled by `xboxpy`?). While I really appreciate the continuous output of nv2a-trace, it does seem like it'd be a reasonable tradeoff if a capture could be done far faster on the xbox itself, storing dumps on the harddrive with separate programs built to process the dumps into nicer formats after the fact.

File: XboxHelper.py (hunk 13 above)
Comment (abaire) [reply XboxHelper.py:22]: > I didn't look into specific registers (and don't have time for it), but I assume nv2a-trace already uses the XQEMU names?

It does now, before this PR it was mostly using hex values directly and where it was using names they were not always aligned with the XQEMU conventions.
 
> I believe the XQEMU header should be using the nvidia names.

Very glad to hear it; I think it's also the easiest source of truth for the average contributor.

File: XboxHelper.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (abaire) [reply XboxHelper.py]: (Sorry, I'm responding to comments in reverse order so will be somewhat brief here to reduce duplication)

The issue I've run into (and other recent users) is a bit the general usability issue of a capture taking very long (I've definitely gone well over two hours for a frame) but more so transient instability that makes it feel like the longer the capture takes, the more likely it is to crash somewhere along the way, requiring an additional attempt.

I don't think it's necessarily a huge deal if everything crashes once the capture is complete (given that there are #fixme's regarding returning to original state I'd worry that a second capture in a session might not be entirely trustable), but it is extremely frustrating if 90 minutes into an attempt there's a failure in xboxpy retrieving memory due to a networking timeout. Most of my motivation in attempting to reduce waits is around the hope that less time spent capturing means less opportunity for the transient failure to manifest coupled with the assumption that having to restart due to a failure 30 minutes in feels less discouraging than the same failure 60 minutes in :)

File: XboxHelper.py (hunk 10 above)
Comment (abaire) [reply XboxHelper.py]: I think this is very useful context and certainly helps me get an idea of where it might make sense to put in further effort. From your other comments below I think it's clear that moving more of the responsibility for tracing onto the hardware would be a good design direction generally. I think it would also reduce the testing surface (i.e., remove worries about behavior dependent on specific xboxpy interfaces) and having the general ability to inject code is obviously useful in a variety of situations, becoming even more useful if the bar is lowered by allowing C in addition to asm.

File: NV2ALog.py (hunk 4 above)
Comment (abaire) [reply NV2ALog.py]: Totally makes sense to me, particularly since it would not be very hard to run the decoder in a separate thread to retain the current sense of progress. I'd be happy to take a stab at this as a followup.

File: Texture.py (hunk 2 above)
Comment (abaire) [reply Texture.py:18]: I have not checked them in nv2a-trace, but I definitely do use the same technique for [the pgraph test suite](https://github.com/abaire/nxdk_pgraph_tests/blob/7b93f25b248c7c01f6ebdd58b60b4f33c8e57813/src/test_host.cpp#L721) and it [appears to work reliably](https://github.com/abaire/nxdk_pgraph_tests_golden_results/wiki/Results-Depth_buffer) (for clarity "CompZ" and "Cy" in those results are with compression enabled).

File: Trace.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (abaire) [reply Trace.py:70]: > If you want to keep it similar, then please remove my name & rephrase it.

Totally understand, attribution has been removed and I did some rewording.

> However, I'd suggest to remove the comment entirely, or look up for a trustable source like nouveau, or run your own experiments and document how hardware behaves (potentially on XboxDevWiki or nouveau / envytools, and then link the corresponding docs from nv2a-trace).

This is the intent of the `#TODO`; once I cycle back to this I'll dig deeper to verify the behavior (I'd be surprised if it turns out to be something other than your previous experience).

> I remember traces taking long for me, but not that long. I usually connect my Xbox directly to my laptop for nv2a-trace though. However, chances are that I was also running a different xboxpy, seeing how many branches of it I have locally. I also remember messing a lot with xbdm and other networking protocols to speed it up. There's at least one commit in [XboxDev/xboxpy#25](https://github.com/XboxDev/xboxpy/pull/25) but I don't know if it helps. There might be more elsehwere.

Cool, I'll look at that next time I'm doing a dump. I'm also testing with a hard wired connection so I doubt network latency attributable to the physical connection has any meaningful impact.

> Unfortunately I've lost track of most xboxpy / nv2a-trace stuff over the years, because nobody else was trying it, so I stopped bothering with pushing or documenting my findings. I digress again, but when I was working on some GPU related demos I would also run the nxdk code on the PC using a modified pbkit / XGU and inject the generated pushbuffers via network. So I'd run the logic on PC, and inject pushbuffers into different emulators and real consoles by broadcasting it via network, then reading the framebuffer remotely to verify emulation / drivers. Most of my Xbox work in the end was automated and done remotely without touching the console. My console was headless most of the time. I might clean up and push these things sometime, but I probably won't have time / interest anytime soon.

This is a fascinating approach and there are some parallel discussions around how to set up a combination of null graphics emulators and GitHub-action driven hardware to do CI on Discord. If you do find the time I think there may be a number of newer contributors to various projects (like myself) who would be extremely interested in catching up on the results of your previous efforts!

File: Trace.py (hunk 12 above)
Comment (abaire) [reply Trace.py]: (marking resolved as this was fixed)

**Comment by abaire** [comment]
Hey @JayFoxRox I believe the other maintainers are looking to you to sign off; apart from the pre-commit and autoformatting where you've asked for other opinions, are there any other outstanding items that need to be resolved in this PR? 

Just to outline the state as I understand it:
- I've added some TODOs on items that need more research and plan to get back to them in the next couple weeks.
- I've removed attribution on the explanatory comments and tried to make it clear that these are areas that need research via said TODOs
- As a short-term followup I'd like to take another step down the path of separating dumping and decoding as you've suggested
- As a longer-term followup I think it'd be extremely useful to move more of the dumping onto the hardware

There's at least one other contributor looking to join in, so I want to make sure this PR isn't blocked on something I've missed and that the proposed followups are acceptable.

**Comment by antangelo** [comment]
I ran into this on my Xbox while running this branch plus a few additional method hooks (so the code should be identical to this branch up to this point):
```
WARNING: real DMA_PUSH_ADDR 0x3DFCCA0 != expected address 0x3DFCCA0. Aborting
Warning: FIFO kick failed
```
I was trying to reproduce on this branch alone, without changes, but xboxpy is crashing almost every time and I don't want to keep continuously restarting my Xbox.

File: kick_fifo.asm [hunk 22]
```diff
@@ -2,49 +2,57 @@
...
+%define CACHE_PUSH_STATE            0xFD003220
+%define DMA_PUSH_ADDR               0xFD003240
+%define DMA_PULL_ADDR               0xFD003244
+
 kick_fifo:
 
-; Load expected_put into reg
-mov edx, dword [esp+4]
+; Load `expected_push` into EDX
+mov edx, dword [ebp+4]
```
Comment (antangelo) [inline kick_fifo.asm:13]: I believe this should be `esp+4`. Probably an artifact of a previous version of this file which had `mov ebp, esp` above. Seems to have resolved the issue I posted about for me.

Comment (abaire) [reply kick_fifo.asm]: Good catch, it is indeed a regression related to the partial revert. Fixed!

**Comment by JayFoxRox** [comment]
> Hey @JayFoxRox I believe the other maintainers are looking to you to sign off; apart from the pre-commit and autoformatting where you've asked for other opinions, are there any other outstanding items that need to be resolved in this PR?

I don't really care what happens to this repo.
I also don't have time to do a more thorough review or resolve all remaining discussions.

As long as I'm not being quoted from the review / none of my details get commited, everything should be good.
I'll trust you and the other maintainers to verify this.

I don't fully remember all the technical details either, so it probably won't take long for you to be the expert on nv2a-trace; I probably can't provide much more input than I already have.
So I'll just trust your changes / future plans and expect the other maintainers to check for trustworthy code (specifically in terms of security, quality and maintainability).

File: README.md (hunk 5 above)
Comment (mborgerson) [reply README.md:57]: I don't think having auto-formatting scripts in this repository is necessary, but I don't object strongly. I also have no objections to using black / following it's style conventions, etc.

File: Texture.py [hunk 23]
```diff
@@ -1,163 +1,326 @@
-#FIXME: Move to xboxpy.NV2A.PGRAPH.Texture
+"""Provides functions for manipulating xbox textures."""
 
-from xboxpy import nv2a
+# pylint: disable=consider-using-f-string
```
Comment (mborgerson) [inline Texture.py:3]: I prefer not to disable these linter warnings globally, and just let the linter complain so this file eventually does get cleaned up

File: Texture.py [hunk 24]
```diff
@@ -1,163 +1,326 @@
...
+        swizzle_unk2=swizzle_unk2,
+        swizzled=swizzled,
+    )
+
+
+def dump_texture(xbox, offset, pitch, fmt_color, width, height):
+    """Convert the texture at the given offset into a PIL.Image."""
+    img = None
+
+    if fmt_color == 0x0:
```
Comment (mborgerson) [inline Texture.py:241]: A dictionary would be preferable over this monstrous if/elif structure

File: Texture.py [hunk 25]
```diff
@@ -1,163 +1,326 @@
...
+            fmt,
+            "float" if is_float else "fixed",
+            "swizzled" if swizzled else "unswizzled",
+        )
+    )
+
+
+def read_texture_parameters(xbox: Xbox) -> TextureParameters:
+    """Reads the current texture state"""
+    color_pitch = xbox.read_u32(0xFD400858)
```
Comment (mborgerson) [inline Texture.py:163]: Would like to these addresses eventually replaced with named constants. Can wait for follow up however

File: README.md (hunk 5 above)
Comment (abaire) [reply README.md:57]: Fair enough; I'm still not sure I understand the reservations around automation, but the pre-commit hook is intentionally opt-in and I hope it'll save anybody who starts to contribute frequently from having to write the same thing manually.

File: Texture.py (hunk 24 above)
Comment (abaire) [reply Texture.py:241]: Agreed, as discussed with @JayFoxRox the entire decoding approach should be refactored. Given that this PR is not introducing this code I'd propose this should be handled in the larger refactoring.

File: README.md (hunk 5 above)
Comment (mborgerson) [reply README.md:57]: I don't dispute the value of automation nor code formatters, I use them regularly.

My argument is against having them run automatically before commit. When I commit something, I want it committed exactly as it is. Also have PTSD from automated tools mucking things up, so I want to be able to easily see and undo what the tool does.

File: Texture.py (hunk 2 above)
Comment (mborgerson) [reply Texture.py:18]: > This is what I do in the pgraph test suite; relatively confident that it works correctly (since it does there, and so far has for all my test cases here), but if there are any known reasons not to take this approach please let me know.

The main reason to avoid reading from this address range is performance
